backend/src/v4/models/mitigation.js

```diff
@@ -132,11 +176,103 @@ class Mitigation {
 			mitigations[i] = newMitigation;
 		}
 
-		await this.clearAll(account);
+		if (clearAll) {
+			await this.clearAll(account);
+		}
+
 		await mitigationColl.insertMany(mitigations);
 
 		return mitigations;
 	}
+
+	async update(account, id, updateData) {
+		await db.updateOne(account, "mitigations", { _id: id }, updateData);
+	}
+
+	async addRiskRefToMitigation(account, mitigation, riskReference) {
+		const referencedRisks = mitigation.referencedRisks;
+
+		if (!referencedRisks.includes(riskReference)) {
+			referencedRisks.push(riskReference);
+			await this.update(account, mitigation._id, { $set: { referencedRisks } });
+		}
+	}
+
+	async updateMitigationsFromRisk(account, model, oldRisk, updatedRisk) {
+		const riskId = updatedRisk._id;
+		const oldStatusIsResolved = !!oldRisk?.mitigation_desc && isMitigationStatusResolved(oldRisk.mitigation_status);
+		const newStatusIsResolved = !!updatedRisk?.mitigation_desc && isMitigationStatusResolved(updatedRisk.mitigation_status);
+
+		// if risk was and remains unresolved
+		if (!oldStatusIsResolved && !newStatusIsResolved) {
+			return;
+		}
+
+		const mitigationDetails = {
+			mitigation_desc: updatedRisk.mitigation_desc,
+			mitigation_detail: updatedRisk.mitigation_detail,
+			mitigation_stage: updatedRisk.mitigation_stage,
+			mitigation_type: updatedRisk.mitigation_type,
+			mitigation_status: updatedRisk.mitigation_status,
+			category: updatedRisk.category,
+			location_desc: updatedRisk.location_desc,
+			element: updatedRisk.element,
+			risk_factor: updatedRisk.risk_factor,
+			scope: updatedRisk.scope,
+			associated_activity: updatedRisk.associated_activity
+		};
+		const formattedReference = formatRiskReference(account, model, riskId);
+
+		// if risk becomes resolved
+		if (!oldStatusIsResolved && newStatusIsResolved) {
+			const mitigations = await this.findMitigationSuggestions(account, mitigationDetails, []);
+			const mitigation = mitigations[0];
+			if (!mitigation) {
+				try {
+					await this.insert(account, [{ ...mitigationDetails, referencedRisks: [formattedReference] }], false);
+				} catch {
+					// do nothing if the mitigation was not inserted
+				}
+			} else if (mitigation.referencedRisks) {
+				await this.addRiskRefToMitigation(account, mitigation, formattedReference);
+			}
+		} else {
+			// if risk was already resolved
+			const allMitigations = await this.findMitigationSuggestions(account, {}, []);
```

Comment:
It's a bit over kill to load in the entire mitigation register and mangually pick them out.

You can do a more specific query like {referecedRisks: formattedReference} so you only pull the relevant entry.

then just do a mongo query to pull the reference if you there are other references https://docs.mongodb.com/manual/reference/operator/update/pull/


Or if you feeling brave - you should be able to do all that in a single mongo pipeline